torch/nested/_internal/ops.py

```diff
@@ -430,25 +430,66 @@ def linear_backward_default(func, *args, **kwargs):
     return (ds, dw, db)
 
 
-@register_jagged_func(torch.ops.aten._to_copy.default, "self: jt_all")
-def to_copy_default(func, *args, **kwargs):
+@register_jagged_func(
+    torch.ops.aten._to_copy.default,
+    "input: jt_all, dtype: any?, layout: any?, device: any?, pin_memory: any?, non_blocking: any?, memory_format: any?",
+)
+def _to_copy_default(func, *args, **kwargs):
     _, new_kwargs = normalize_function(
         func, args=args, kwargs=kwargs, normalize_to_only_use_kwargs=True
     )
 
-    inp = new_kwargs.pop("input")
-    # don't change layout
-    new_kwargs.pop("layout")
+    inp: NestedTensor = new_kwargs.pop("input")
+    new_layout = new_kwargs.pop("layout")
+    if new_layout is None:
+        new_layout = inp.layout
+
+    if new_layout not in [torch.strided, torch.jagged]:
+        raise ValueError("Nested Tensors can only have jagged and strided layouts")
 
     new_values = func(inp._values, **new_kwargs)
-    # NB: Purposefully keep offsets on the old device.
-    return NestedTensor(new_values, **extract_kwargs(inp))
+
+    if new_layout == torch.jagged:
+        # Copy to a new Python subclass NestedTensor
+        return NestedTensor(
+            new_values,
+            lengths=inp.lengths(),
+            **extract_kwargs(inp),
+        )
+
+    # Create a new C++ NT from the Python NestedTensor
+    # Start by creating metadata needed by C++ NT
+    ragged_source = inp.lengths() if inp.lengths() is not None else inp.offsets().diff()
+    nested_sizes = torch.empty(
+        (inp.offsets().shape[0] - 1, new_values.dim()), dtype=torch.int64
+    )
+    non_ragged_dims = list(range(new_values.dim()))
+    non_ragged_dims = (
+        non_ragged_dims[: inp._ragged_idx - 1] + non_ragged_dims[inp._ragged_idx :]
+    )
+    nested_sizes[:, non_ragged_dims] = torch.tensor(
+        inp._size[1 : inp._ragged_idx] + inp._size[inp._ragged_idx + 1 :]
+    )
+    nested_sizes[:, inp._ragged_idx - 1] = ragged_source
+    nested_strides = torch.empty_like(nested_sizes)
+    nested_strides[:, :] = torch.tensor(inp._strides[1:])
```

Comment:
idk how this compares perf-wise, but alternatively you could just `unbind()` the jagged NT and use the constituent shapes strides to construct `nested_sizes` / `nested_strides`. it'd be simpler although we do want what's faster

Reply:
I can benchmark when the errors are fixed

Reply:
the current code runs in 0.25ms for the example in the test case. The easier code:
```
    constituents = inp.unbind()
    nested_sizes_list = [list(c.shape) for c in constituents]
    nested_sizes = torch.tensor(nested_sizes_list, dtype=torch.int64)
    nested_strides_list = [list(c.stride()) for c in constituents]
    nested_strides = torch.tensor(nested_strides_list, dtype=torch.int64)
    nested_offsets = inp.offsets() * functools.reduce(
        lambda a, b: a * b, inp._values.shape[1:]
    )
    nested_offsets = nested_offsets[:-1]
```
runs in 0.67ms for the example in the test case, so about 2.7x slower.

Reply:
Thanks for checking!